src/ophyd_async/core/device.py

```diff
@@ -140,40 +149,52 @@ async def __aenter__(self) -> "DeviceCollector":
 
     async def _on_exit(self) -> None:
         # Name and kick off connect for devices
-        tasks: Dict[asyncio.Task, str] = {}
+        tasks: Dict[str, asyncio.Task] = {}
         for name, obj in self._objects_on_exit.items():
             if name not in self._names_on_enter and isinstance(obj, Device):
                 if self._set_name and not obj.name:
                     obj.set_name(name)
                 if self._connect:
-                    task = asyncio.create_task(obj.connect(self._sim))
-                    tasks[task] = name
+                    tasks[name] = asyncio.create_task(
+                        obj.connect(self._sim, timeout=self._timeout)
+                    )
         # Wait for all the signals to have finished
         if tasks:
             await self._wait_for_tasks(tasks)
 
-    async def _wait_for_tasks(self, tasks: Dict[asyncio.Task, str]):
-        done, pending = await asyncio.wait(tasks, timeout=self._timeout)
-        if pending:
-            msg = f"{len(pending)} Devices did not connect:"
-            for t in pending:
-                t.cancel()
-                with suppress(Exception):
-                    await t
-                e = t.exception()
-                msg += f"\n  {tasks[t]}: {type(e).__name__}"
-                lines = str(e).splitlines()
+    async def _wait_for_tasks(self, tasks: Dict[str, asyncio.Task]):
+        names = tasks.keys()
+        results = await asyncio.gather(*tasks.values(), return_exceptions=True)
+
+        disconnected = 0
+        failed: Dict[str, Exception] = {}
+
+        msg = ""
+
+        for name, result in zip(names, results):
+            if result and isinstance(result, ConnectionTimeoutError):
+                msg += f"\n  {name}: {type(result).__name__}"
+                lines = str(result).splitlines()
                 if len(lines) <= 1:
-                    msg += f": {e}"
+                    msg += f": {result}"
                 else:
                     msg += "".join(f"\n    {line}" for line in lines)
+
+                disconnected += 1
+
+            elif result and isinstance(result, Exception):
+                failed[name] = result
+
+        if disconnected:
+            msg = f"{disconnected} Devices did not connect:" + msg
             logging.error(msg)
-        raised = [t for t in done if t.exception()]
-        if raised:
-            logging.error(f"{len(raised)} Devices raised an error:")
-            for t in raised:
-                logging.exception(f"  {tasks[t]}:", exc_info=t.exception())
-        if pending or raised:
+
+        if failed:
+            logging.error(f"{len(failed)} Devices raised an error:")
+            for name, exception in failed.items():
+                logging.exception(f"  {name}:", exc_info=exception)
+
+        if disconnected or failed:
```

Comment:
Had a go at a slightly more concise version of this, which I like because it handles timeouts and errors in the same way, what do you think?

```python
    async def _wait_for_tasks(self, tasks: Dict[str, asyncio.Task]):
        names = tasks.keys()
        results = await asyncio.gather(*tasks.values(), return_exceptions=True)

        timed_out: Dict[str, ConnectionTimeoutError] = {}
        failed: Dict[str, Exception] = {}

        for name, result in zip(names, results):
            if result and isinstance(result, ConnectionTimeoutError):
                timed_out[name] = result
            elif result and isinstance(result, Exception):
                failed[name] = result

        if len(timed_out) > 0:
            logging.error(f"{len(timed_out)} devices timed out during connection:")
            for name, exception in timed_out.items():
                logging.exception(f"  {name}:", exc_info=exception)
        if len(failed) > 0:
            logging.error(f"{len(failed)} devices raised an error:")
            for name, exception in failed.items():
                logging.exception(f"  {name}:", exc_info=exception)

        if len(timed_out) > 0 or len(failed) > 0:
            raise NotConnected("Not all Devices connected")

```

At this point the only purpose of having separate handling code for `ConnectTimeoutError` and other kinds of exceptions is to sort them, i.e. so we can have

```shell
ERROR: 2 Devices timed out
device_a: <taceback>...ConnectionTimeOutError
device_b: <taceback>...ConnectionTimeOutError

ERROR: 2 Devices errored
device_c: <taceback>...IOError
device_d: <taceback>...PVaccessError
```

So as a separate question: Do we care any more, they are all just tracebacks at the end of the day, does it matter if it says something like this?

```shell
ERROR: 4 Devices failed to
device_a: <taceback>...ConnectionTimeOutError
device_c: <taceback>...IOError
device_b: <taceback>...ConnectionTimeOutError
device_d: <taceback>...PVaccessError
```

...If not then we could simply further and just say "if exception: log it"

Reply:
I don't think reporting hundreds of the same tracebacks when the IOC is down is very helpful, we should think of a way to squash it. My suggestion:
- If a `Signal.connect` sees the PV is disconnected, it raises `NotConnected([self.source])` from the original error
- `Device.connect` aggregates all errors into a single `NotConnected(indented(errors))`, and `log.exception`s any non-NotConnected errors
- `DeviceCollector` does the same error handling as `Device.connect`

Reply:
Ok, how about this, log `NotConnected` at debug level, log all other exceptions at error level

Reply:
I'm happy with that as long as it is clearly documented!